great_expectations/datasource/data_connector/configured_asset_azure_data_connector.py

```diff
@@ -165,21 +165,22 @@ def _get_full_file_path_for_asset(
         # asset isn't used in this method.
         # It's only kept for compatibility with parent methods.
         # Pandas and Spark execution engines utilize separate path formats for accessing Azure Blob Storage service.
-        full_path: str
-        if isinstance(self.execution_engine, PandasExecutionEngine):
-            full_path = os.path.join(
-                f"{self._account_name}.blob.core.windows.net", self._container, path
-            )
-        elif isinstance(self.execution_engine, SparkDFExecutionEngine):
-            full_path = os.path.join(
-                f"{self._container}@{self._account_name}.blob.core.windows.net", path
-            )
-            full_path = f"wasbs://{full_path}"
-        else:
+        if not hasattr(
+            self.execution_engine, "get_azure_blob_storage_object_url_template"
+        ):
             raise ge_exceptions.DataConnectorError(
                 f"""Illegal ExecutionEngine type "{str(type(self.execution_engine))}" used in \
 "{self.__class__.__name__}".
 """
             )
 
-        return full_path
+        template_arguments: dict = {
+            "account_name": self._account_name,
+            "container": self._container,
+            "path": path,
+        }
+
+        # noinspection PyUnresolvedReferences
+        return self.execution_engine.get_azure_blob_storage_object_url_template(
```

Comment:
The implication here is that if someone introduces a new ExecutionEngine, they also have to know to add a bunch of static methods to that engine to support specific DataConnectors can call back to the execution engine for details on how to format paths specific to that data connector. I feel this new approach is more convoluted than the old one, where the DataConnector had conditional logic per ExecutionEngine

What I propose is a registered list of `DataConnectorResolvers` in some kind of `DataConnectorResolverRegistry`. If we want to dispense with types, this registry could be a `dict[tuple[EngineType, ConnectorType], lambda]` that exists globally (in the DataSource, I suppose), or it could be a property of the `ExecutionEngine` as `dict[ConnectorType, lambda]`. The implementation of `_get_full_file_path_for_asset` could call `self.execution_engine.resolve_path(self.__class__.__name__, dict)` which would then retrieve the lambda and invoke it.

Benefits of this approach:

- No magic static methods we need to detect with `hasattr`
- A 3rd party ExecutionEngine could add resolvers for the DataConnectors it wants to support, and it would be obvious where and how to do it (the more I think about it, the more I believe this registry should live at the DataSource level)
- Alternatively/in addition, 3rd party DataConnectors could register the execution engines they want to support
- The `ExecutionEngine.resolve_path` method could be defined at the highest level and the `DataConnectorError` would be raised there if the necessary key in the dictionary was missing.
- From a design standpoint, it's kind of like a join table in SQL, or a flavor of the strategy pattern

Points for further consideration:

- Getting paths only matters for engines and connectors that support pathing. There is probably some opportunity for specialization here to exclude non-path repos (like SQL). Maybe this is where thinking at the DataSource level (and specializing at that level) makes sense, since it has knowledge of both engines and connectors
- For our own (first party) engines and connectors, we should decide where this registry gets populated. There is likely an obvious initialization routine for GE where this should live.

Reply:
@donaldheppner There is so much good in your idea on so many levels!  Thank you for suggesting it.  I implemented it (please re-review).  The only couple of (minor) things I had to do somewhat differently was:
* Use string names of the `DataConnector` and `ExecutionEngine` objects in the registry's `Tuple` key instead of the object types themselves in order to avoid circular dependencies in `Python` (this will no longer be an issue in `Python-4`).
* Create an intermediate mapping from `DataConnector` names to cloud storage names so as to avoid registry key duplication (we have two `DataConnector` types, Inferred and Configured, for each cloud storage type.

Thank you so much!!!

Reply:
Awesome, looks great Alex! You captured the essence of the idea perfectly